hoomd/hpmc/pytest/test_shape_updater.py

```diff
@@ -0,0 +1,246 @@
+
+
+@pytest.mark.parametrize("shape_move_class,attr,value", shape_move_valid_attrs)
+def test_valid_setattr_shape_move(shape_move_class, attr, value):
+    """Test that the shape move classes can get and set attributes."""
+
+    move = shape_move_class()
+
+    setattr(move, attr, value)
+    assert getattr(move, attr) == value
+
+
+@pytest.mark.parametrize("attr,value", shape_updater_valid_attrs)
+def test_valid_setattr_shape_updater(attr, value):
+    """Test that the Shape updater can get and set attributes."""
+
+    updater = hpmc.update.Shape(trigger=1)
+
+    setattr(updater, attr, value)
+    assert getattr(updater, attr) == value
+
+
+@pytest.mark.parametrize("obj,attr,value", type_parameters)
+def test_type_parameters(obj, attr, value):
+    getattr(obj, attr)["A"] = value
+    assert getattr(obj, attr)["A"] == value
+
+
+def test_vertex_shape_move(device, simulation_factory,
+                           two_particle_snapshot_factory):
+
+    move = VertexShapeMove()
+    move.volume["A"] = 1
+
+    updater = hpmc.update.Shape(trigger=1, step_size=0.2, nsweeps=2)
+    updater.shape_move = move
+
+    mc = hoomd.hpmc.integrate.ConvexPolyhedron()
+    mc.d["A"] = 0
+    mc.a["A"] = 0
+    mc.shape["A"] = dict(vertices=verts)
+
+    # create simulation & attach objects
+    sim = simulation_factory(two_particle_snapshot_factory(d=10))
+    sim.operations.integrator = mc
+    sim.operations += updater
+
+    # test attachmet before first run
+    assert not move._attached
+    assert not updater._attached
+
+    sim.run(0)
+
+    # test attachmet after first run
+    assert move._attached
+    assert updater._attached
+
+    # run with 0 probability of performing a move:
+    #  - shape should remain unchanged
+    #  - all moves accepted
+    move.move_probability = 0
+    sim.run(20)
+    assert np.allclose(mc.shape["A"]["vertices"], verts)
+    print(updater.shape_moves)
+    assert updater.shape_moves[0] == 40
+    assert updater.shape_moves[1] == 0
+
+    # always attempt a shape move:
+    #  - shape should change, some attemps are accepted but
+    #  - volume should remain unchanged
+    move.move_probability = 1
+    sim.run(20)
+    assert updater.shape_moves[0] != 0
+    assert updater.shape_moves[1] != 0
+    assert np.sum(updater.shape_moves) == 40
+    assert not np.allclose(mc.shape["A"]["vertices"], verts)
+    assert np.isclose(updater.total_particle_volume, 2)
+
+
+def test_python_callback_shape_move(device, simulation_factory,
+                                    two_particle_snapshot_factory):
+
+    class ScaleEllipsoid:
+        """Toy class to randomly squashes a sphere into oblate ellipsoid with
+           constant volume.
+        """
+
+        def __init__(self, a, b, c):
+            self.vol_factor = 4 * np.pi / 3
+            self.volume = self.vol_factor * a * b * c
+            self.default_dict = dict(ignore_statistics=True)
+
+        def __call__(self, type_id, param_list):
```

Comment:
See there is no reason to be passed the `param_list` as this could just exist in the class instance. Only the `type_id` is necessary.

Reply:
In the python shape move, `param_list` refers to the list of parameters that completely determine your shape through some arbitrary user defined parametrization. The callback function only returns a shape definition dictionary based on the input parameters, but the actual trial moves and parameters are performed and stored in C++ respectively. Because of this, the callback function requires  `param_list` as an argument.

The most common use case is to perform trial moves on the truncation degree of truncated tetrahedra where the callback would get the truncation degree as an input and return the vertices corresponding to that truncation.

Reply:
Okay, so I am starting to understand I think, then is the code for the Python shape move just to randomly perturb the `param_list` and get what that perturbation results in, in C++? If so why take this approach over giving complete control to Python so users can just define the shape move there?

Reply:
Yes, the trial moves are performed in C++, you only get the shape definition from python. I agree that performing the shape move completely in python would be more beneficial for the user in terms of flexibility and would also better match the current name. However, this is the original implementation and we didn't venture into changing it. If we decide to go that route I suggest we do it in a different PR after this one has been merged and released.